server/server.go

```diff
@@ -102,35 +100,46 @@ type Server struct {
 }
 
 // NewServer creates a Server from a server.Context.
-func NewServer(ctx Context, stopper *stop.Stopper) (*Server, error) {
-	if _, err := net.ResolveTCPAddr("tcp", ctx.Addr); err != nil {
-		return nil, errors.Errorf("unable to resolve RPC address %q: %v", ctx.Addr, err)
+func NewServer(srvCtx Context, stopper *stop.Stopper) (*Server, error) {
+	if _, err := net.ResolveTCPAddr("tcp", srvCtx.Addr); err != nil {
+		return nil, errors.Errorf("unable to resolve RPC address %q: %v", srvCtx.Addr, err)
 	}
 
-	if ctx.Insecure {
-		log.Warning(context.TODO(), "running in insecure mode, this is strongly discouraged. See --insecure.")
+	if srvCtx.Ctx == nil {
+		// TODO(radu): we should make sure the callers set this explicitly instead.
+		srvCtx.Ctx = context.Background()
+	}
+	tracer := tracing.TracerFromCtx(srvCtx.Ctx)
+	if tracer == nil {
+		tracer = tracing.NewTracer()
+		srvCtx.Ctx = tracing.WithTracer(srvCtx.Ctx, tracer)
+	}
+
+	ctx := srvCtx.Ctx
```

Comment:
This kind of adds to the confusion here. Why not just always references srvCtx.Ctx where you need this instead of creating a new local?